external/kafka-0-10-token-provider/src/main/scala/org/apache/spark/kafka010/KafkaTokenUtil.scala

```diff
@@ -222,14 +230,25 @@ private[spark] object KafkaTokenUtil extends Logging {
     }
   }
 
-  def isTokenAvailable(): Boolean = {
-    UserGroupInformation.getCurrentUser().getCredentials.getToken(
-      KafkaTokenUtil.TOKEN_SERVICE) != null
+  def findMatchingToken(sparkConf: SparkConf,
+      bootStrapServers: String): Option[KafkaTokenClusterConf] = {
+    val tokens = UserGroupInformation.getCurrentUser().getCredentials.getAllTokens.asScala
+    val clusterConfigs = tokens
+      .filter(_.getService().toString().startsWith(TOKEN_SERVICE_PREFIX))
+      .map { token =>
+        KafkaTokenSparkConf.getClusterConfig(sparkConf, getClusterIdentifier(token.getService()))
+      }
+      .filter { clusterConfig =>
+        Pattern.compile(clusterConfig.targetServersRegex).matcher(bootStrapServers).matches()
```

Comment:
What does `bootStrapServers` look like? Is it a single address, or like a list of addresses e.g. separated by commas?

I think in the latter case you should test each address against the regex, otherwise it becomes very, very hard to write a regex for matching all possible permutations of how to write that list.

Reply:
The mentioned config format is the following: `host1.company.com:9092,host2.company.com:9092`.

Such case one can write for ex. the following regex: `.*company\.com:9092`.

Not yet see in which situation is it easier to write regex? If a substring of a string matches with a regex then the whole string should match, shouldn't it? If you could give an example would help.

Reply:
That's a very simplistic case. In that case the `.*` is matching  "host1.company.com:9092,host2.", and the rest is matching the end of the config.

I think it's much better, easier to understand, and less prone to errors to have the regex match individual addresses instead of the whole string. It's just a matter of calling `Utils.stringToSeq` before applying the regex.

Reply:
Makes sense, the actual implementation can end-up in malformed regex. Changed.